src/Features/LanguageServer/Protocol/ExternalAccess/VSCode/API/VSCodeTelemetryLogger.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.CodeAnalysis.Internal.Log;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.ExternalAccess.VSCode.API;
+
+/// <summary>
+/// Allows VSCode to implement a telemetry logger for <see cref="ILogger"/> events.
+/// </summary>
+internal abstract class VSCodeTelemetryLogger : ILogger
+{
+    private readonly Dictionary<FunctionId, string> _functionIdToString = new Dictionary<FunctionId, string>();
```

Comment:
❗ `Dictionary<TKey, TValue>` isn't thread-safe, so either pre-compute all the values in a static constructor, or use a `lock` inside `GetFunctionIdAsString`.

Reply:
It might make sense to create a helper class for this since `TraceSourceLogger` needs it as well.

Reply:
Added helper extension method to convert functionids to strings and used it for TraceSourceLogger as well